src/audio/module_adapter/library/userspace_proxy.c

```diff
@@ -44,6 +49,139 @@ DECLARE_TR_CTX(userspace_proxy_tr, SOF_UUID(userspace_proxy_uuid), LOG_LEVEL_INF
 
 static const struct module_interface userspace_proxy_interface;
 
+struct user_worker {
+	k_tid_t thread_id;			/* ipc worker thread ID			*/
+	uint32_t reference_count;		/* module reference count		*/
+	void *stack_ptr;			/* pointer to worker stack		*/
+	struct k_work_user_q work_queue;
+	struct k_event event;
+};
+
+static struct user_worker worker;
+
+static int user_worker_create(void)
+{
+	if (worker.reference_count) {
+		worker.reference_count++;
+		return 0;
+	}
+
+	worker.stack_ptr = user_stack_allocate(CONFIG_SOF_STACK_SIZE, K_USER);
+	if (!worker.stack_ptr) {
+		tr_err(&userspace_proxy_tr, "Userspace worker stack allocation failed.");
+		return -ENOMEM;
+	}
+
+	k_event_init(&worker.event);
+	k_work_user_queue_start(&worker.work_queue, worker.stack_ptr, CONFIG_SOF_STACK_SIZE, 0,
+				NULL);
+
+	worker.thread_id = k_work_user_queue_thread_get(&worker.work_queue);
+
+	k_thread_access_grant(worker.thread_id, &worker.event);
+
+	worker.reference_count++;
+	return 0;
+}
+
+static void user_worker_free(void)
+{
+	/* Module removed so decrement counter */
+	worker.reference_count--;
+
+	/* Free worker resources if no more active user space modules */
+	if (worker.reference_count == 0) {
+		k_thread_abort(worker.thread_id);
+		user_stack_free(worker.stack_ptr);
+	}
+}
+
+static int user_work_item_init(struct userspace_context *user_ctx, struct k_heap *user_heap)
+{
+	struct user_work_item *work_item = NULL;
+	int ret;
+
+	ret = user_worker_create();
+	if (ret)
+		return ret;
+
+	work_item = sof_heap_alloc(user_heap, SOF_MEM_FLAG_COHERENT, sizeof(*work_item), 0);
```

Comment:
So I assume we have 1 worker for all cores given we have coherent memory, if so probably best to call this out in the comments.